x-pack/plugins/security_solution/server/endpoint/routes/trusted_apps/service.ts

```diff
@@ -27,7 +29,38 @@ import {
   osFromExceptionItem,
   updatedTrustedAppToUpdateExceptionListItemOptions,
 } from './mapping';
-import { TrustedAppNotFoundError, TrustedAppVersionConflictError } from './errors';
+import {
+  TrustedAppNotFoundError,
+  TrustedAppVersionConflictError,
+  TrustedAppPolicyNotExistsError,
+} from './errors';
+import { PackagePolicyServiceInterface } from '../../../../../fleet/server';
+import { PackagePolicy } from '../../../../../fleet/common';
+
+const getUnexistingPoliciesFromTrustedApp = async (
+  savedObjectClient: SavedObjectsClientContract,
+  packagePolicyClient: PackagePolicyServiceInterface,
+  trustedApp: PutTrustedAppUpdateRequest | PostTrustedAppCreateRequest
+): Promise<PackagePolicy[]> => {
+  if (
+    !trustedApp.effectScope ||
+    trustedApp.effectScope.type === 'global' ||
+    (trustedApp.effectScope.type === 'policy' && isEmpty(trustedApp.effectScope.policies))
+  ) {
+    return [];
+  }
+
+  const policies = await packagePolicyClient.getByIDs(
+    savedObjectClient,
+    trustedApp.effectScope.policies
+  );
+
+  if (!policies) {
+    return [];
+  }
+
+  return policies.filter((policy) => policy.version === undefined);
```

Comment:
Trying to figure out if this looks correct.  The purpose of this function is to return the IDs of the polices that no longer exist (right?). The `getByIDs()` method of fleet service seems to just search for these IDs and only return what it finds (right? or does it return an array of equal size to input, but some element are null?)

If that is the case, then shouldn't you be using the `trustedApp.effectScope.policies` as the array you iterate over and check that the policy id was returned from Fleet? Like:

```typescript
const foundPolicyIds = policies.map(policy => policy.id);
return trustedApp.effectScope.policies.filter(policy => !foundPolicyIds.includes(policy.id))
````

Reply:
getByIds it's returning the policy or an object with a version key eq to undefined and the id. Example:

```
[
  {
    id: 'fa2b9240-99d2-4bc3-b0bc-6e7caf417154',
    version: 'WzE0NjAxNiw0XQ==',
    name: 'Full protection',
    description: '',
    namespace: 'default',
    policy_id: 'a5507720-e86f-11eb-a87f-51e16104076a',
    enabled: true,
    output_id: '',
    inputs: [ [Object] ],
    package: { name: 'endpoint', title: 'Endpoint Security', version: '0.20.2' },
    revision: 4,
    created_at: '2021-07-21T10:35:18.879Z',
    created_by: 'elastic',
    updated_at: '2021-07-21T10:53:31.068Z',
    updated_by: 'system'
  },
  { id: '9af4a819-dedd-400e-941b-3b2b710ea3c1', version: undefined }
]
```

This is a response when getByIds `9af4a819-dedd-400e-941b-3b2b710ea3c1` and `fa2b9240-99d2-4bc3-b0bc-6e7caf417154`